xwiki-pro-macros-api/src/main/java/com/xwiki/macros/internal/updateReferences/AbstractReferenceUpdateMacroRefactoring.java

```diff
@@ -0,0 +1,146 @@
+
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> compactWikiEntityReferenceSerializer;
+
+    @Inject
+    @Named("macro")
+    private EntityReferenceResolver<String> macroEntityReferenceResolver;
+
+    /**
+     * Returns the list of parameter names in the macro that should be updated during reference refactoring.
+     * @return the list of parameter names to update
+     */
+    public abstract List<String> getParametersToUpdate();
+
+    @Override
+    public Optional<MacroBlock> replaceReference(MacroBlock macroBlock, DocumentReference currentDocumentReference,
+        DocumentReference sourceReference, DocumentReference targetReference, boolean relative)
+        throws MacroRefactoringException
+    {
+        return refactorMacroBlock(macroBlock, currentDocumentReference, sourceReference, targetReference);
+    }
+
+    @Override
+    public Optional<MacroBlock> replaceReference(MacroBlock macroBlock, DocumentReference currentDocumentReference,
+        AttachmentReference sourceReference, AttachmentReference targetReference, boolean relative)
+    {
+        return refactorMacroBlock(macroBlock, currentDocumentReference, sourceReference, targetReference);
+    }
+
+    /**
+     * Refactors a macro block by updating its parameters when they match a given source reference.
+     *
+     * @param <T> the type of entity reference
+     * @param macroBlock the original macro block
+     * @param currentDocumentReference the reference of the document where the macro resides
+     * @param sourceReference the reference to be replaced
+     * @param targetReference the reference that replaces the source
+     * @return an optional containing the updated macro block, or empty if no parameter needed refactoring
+     */
+    public <T extends EntityReference> Optional<MacroBlock> refactorMacroBlock(MacroBlock macroBlock,
+        DocumentReference currentDocumentReference, T sourceReference, T targetReference)
+    {
+        List<String> parametersToUpdate = getParametersToUpdate();
+        if (!parametersToUpdate.isEmpty()) {
+            MacroBlock newMacroBlock = (MacroBlock) macroBlock.clone();
+
+            for (String parameterToUpdate : parametersToUpdate) {
+                String stringMacroReference = macroBlock.getParameter(parameterToUpdate);
+                EntityReference macroReference =
+                    this.macroEntityReferenceResolver.resolve(stringMacroReference, EntityType.DOCUMENT, macroBlock,
+                        sourceReference);
+
+                boolean resolvedRelative = !isReferenceAbsolute(stringMacroReference, macroReference);
+
+                if (macroReference.equals(sourceReference)) {
+                    newMacroBlock.setParameter(parameterToUpdate,
+                        serializeTargetReference(targetReference, currentDocumentReference, resolvedRelative));
+                }
+            }
+            return Optional.of(newMacroBlock);
```

Comment:
I'm not sure but I think it would be better to only return a non-empty result when the macro block has actually been changed.